File: moderate/double_squares.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"math"
	"os"
)

func sqrt(a int) int {
	return int(math.Sqrt(float64(a)))
}

func main() {
	sq := make(map[int]bool)
	for i := 0; i < 46341; i++ {
		sq[i*i] = true
	}

	data, err := os.Open(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}
	defer data.Close()
	reader := bufio.NewReader(data)
	s, err := reader.ReadString('\n')
	if err != nil {
		log.Fatal(err)
	}
	var n int
	fmt.Sscan(s, &n)
	for i := 0; i < n; i++ {
		s, err := reader.ReadString('\n')
		if err != nil {
			break
		}
		var x int
		fmt.Sscan(s, &x)
		num, l, bot, top := 0, make(map[int]bool), sqrt(x/2), sqrt(x)+1
		for j := bot; j < top; j++ {
			t := x - j*j
			if sq[t] && !l[t] {
				l[j*j], num = true, num+1
			}
		}
		fmt.Println(num)
	}
}

File: moderate/double_squares.rb
sq, n = Hash.new, 0
(0..46340).each do |i| sq[i*i] = true end

File.open(ARGV[0]).each_line do |line|
  if n == 0
    n = line.to_i
  else
    x = line.to_i
    num, l, bot, top = 0, Hash.new, Math.sqrt(x/2).to_i, Math.sqrt(x).to_i
    (bot..top).each do |i|
      t = x - i*i
      l[i*i], num = true, num+1 if sq[t] && !l[t]
    end
    puts num
  end
end
